### ISSUE TITLE
How to trian a model to swap faces for arbitrary person?

### ISSUE THREAD

**Issue by ypflll** [deepfakes/faceswap-playground#144]
I am trying to develop a program to swap face for two person like this:
![image](https://user-images.githubusercontent.com/12528364/39347525-d228a628-4a25-11e8-8d31-3e20daeb2846.png)

I tried this repo.
However, I find that your demo just shows how to train a model to swap two fixed persons, namely Trump and Cage. So, if I want to swap faces with Trump and Hillary Clinton, I have to train another model.
If this repo can be used to train a model, which takes pictures of two person (no matter who are they), and swap their faces? And how?

**Comment by Kirin-kun** [comment]
This repo can be used to train any face swapping. You can start with a pre-trained model, like Cage/Trump or you can start with an empty directory. It will fill it for you.

It just takes more time if the two faces are not the same than the ones the model was trained for originally.

Let's say you have the original model that puts the face of Cage on Trump. 

Simplest case: you want to put the face of Cage on Hillary. 

As the model already learned a lot about Cage, you will just need to train the network a bit with the face of Hillary in lieu of the face of Trump and it will learn to swap them faster than if you started with an empty model. Just a few hours, or even less and the loss will go down really fast.

As long as the network learned to swap one of the faces already, to learn to swap it on another face just takes a bit of training. Whether it's A->B or B->A.

Now, if you want to swap two totally unrelated faces, you can re-use the same model files, but it will first have to "forget" about the two original faces and learn the two new faces. It will probably go a little faster, because the poses might be similar, but if it didn't forget enough, you might get features of the original faces on your target.

So yeah, this repo can be used to swap any pair of faces. Starting with an empty model will avoid surprises at conversion time if you're not patient enough.

**Comment by ypflll** [comment]
Thanks for your reply.
So, it is clear that one model just can swap one pair of person. If I want to swap another one's face, a new model need to be trained. I think this is because the fact that Networks learns the feature of a face, not the way to how to swap, so for a new person, feature must be learned again.

I tried a more 'traditional' way to swap two faces, which can get a 'universal' model:
face detection->land mark detection->Delaunay Triangulation->Affine warp and bending.
This model can be used to any pair of face swap with no extra training.  However, the result is far from satisfied.
Do you think I can get a better 'universal' model for this issue?@Kirin-kun

**Comment by Kirin-kun** [comment]
One trained model can put one face (ex Cage) on many persons. But as it has been trained on Trump's poses and features, if you directly try to put it on another person, your result will not be optimal for the poses/features that it never learned. 

You can experiment easily. Just extract another person's face from a set of pictures and convert to them without training. It will work nicely for some of them, but not all of them, like it did for the original pairing.

Your "traditional" swapping doesn't learn any of the features (eyes open/closed, mouth open/closed) of the target face, so it will be fully manual to pair the mouth and eyes and an impossible task.

As for an universal model, what do you mean? Something that wouldn't need training? That's not the point of this repo, which is to showcase an application of deep learning and get people to learn more about it.

**Comment by ypflll** [comment]
By universal I mean, as you mentioned, put Trump's face on many people's picture using one model.
![image](https://user-images.githubusercontent.com/12528364/39357711-39699b24-4a46-11e8-9039-84040a54179e.png)

Or, change Trump's picture with many other people's face.
![image](https://user-images.githubusercontent.com/12528364/39357723-41f97476-4a46-11e8-908a-d9e57b7cee73.png)

As you mentioned, it can be done, but result will not be optimal.
How about I train the model like this: INPUT_A with Trump's picture and INPUT_B with a lot of other people's pictures?

**Comment by Kirin-kun** [comment]
I tried this in the past, as did other people, but I didn't have overall good result. 

The model learns **both** faces. Not just how to warp a face to any face. So if you feed it with multiple faces in B, it will average the features and it will look odd in some cases.

For doing what you want, you'd have to tweak the model and the loss function and that's beyond my knowledge.

**Comment by ypflll** [comment]
@Kirin-kun Thanks for your experience. Great help to me

Any one has tried this is welcomed to report the result. I think I will try this if I have enough time later.
